### PR TITLE
support for nix flake builds

### PR DESCRIPTION
**Pull request by smunix**
add support for nix flake

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +0,0 @@
-[submodule "souffle"]
```
Comment (luc-tielen) [inline .gitmodules:1]: Why is this file removed? I need the submodule to easily update souffle headers every release.

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,98 @@
+{
+  description = "Souffle Datalog bindings for Haskell";
+  inputs = {
+    np.url = "github:nixos/nixpkgs?ref=haskell-updates";
```
Comment (luc-tielen) [inline flake.nix:4]: Can we pin this to a commit? If we always fetch from an upstream channel, it's not deterministic over time.

Comment (luc-tielen) [reply flake.nix:4]: You could take a look at current commit of haskell-updates branch and put that here maybe?

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    hls.url = "github:haskell/haskell-language-server?ref=master";
+  };
+  outputs = { self, np, fu, hls }:
+    with fu.lib;
+    eachSystem [ "x86_64-linux" ] (system:
+      let
+        version = with np.lib;
+          "${substring 0 8 self.lastModifiedDate}.${self.shortRev or "dirty"}";
+        config = { };
+        mkSouffle = { p, gcc ? 10 }:
```
Comment (luc-tielen) [inline flake.nix:15]: Move this function to a separate souffle.nix file and import it here? Or is this not support by nix flakes?

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            souffle-haskell = with haskell.lib;
+              ((addBuildDepends
+                (addBuildTools (callCabal2nix "souffle-haskell" ./. { }) [
+                  hpack
+                  souffle
+                ]) [ makeWrapper ]).overrideAttrs (o: {
+                  version = "${o.version}.${version}";
+                  postPatch = ''
+                    substituteInPlace package.yaml \
+                      --replace "tests:" "executables:"
```
Comment (luc-tielen) [inline flake.nix:69]: What does this do? Disable tests?

File: flake.nix [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                    ${hpack}/bin/hpack -f
+                  '';
+                  postFixup = ''
+                    wrapProgram $out/bin/souffle-haskell-test \
```
Comment (luc-tielen) [inline flake.nix:73]: Nice!

File: lib/Language/Souffle/Interpreted.hs [hunk 6]
```diff
@@ -367,7 +367,7 @@ datalogProgramFile prog datalogDir = do
 
 locateSouffle :: IO (Maybe FilePath)
 locateSouffle = do
-  let locateCmd = (shell "which souffle") { std_out = CreatePipe }
+  let locateCmd = (proc "souffle" []) { std_out = CreatePipe }
```
Comment (luc-tielen) [inline lib/Language/Souffle/Interpreted.hs:370]: I don't think this is correct. Before we searched for souffle, now you are directly invoking it, which is wrong. I may need to add a test specifically for this later..

Can you revert this please?

**Review by luc-tielen — CHANGES_REQUESTED** [review]
Overall looks good. But left some remarks on things that need to be changed/fixed. Thanks for doing this!

File: flake.nix (hunk 3 above)
Comment (luc-tielen) [reply flake.nix:15]: Ideally, we'd upstream this to nixpkgs, so we don't have to specify it here, but with each new souffle release we would lag behind immediately.

File: flake.nix (hunk 2 above)
Comment (smunix) [reply flake.nix:4]: This is actually pinned (see `flake.lock`)

Comment (luc-tielen) [reply flake.nix:4]: Ah yes I see it now, my bad!

File: lib/Language/Souffle/Interpreted.hs (hunk 6 above)
Comment (smunix) [reply lib/Language/Souffle/Interpreted.hs:370]: sorry, I had overlooked this. reverting!

File: .gitmodules (hunk 1 above)
Comment (smunix) [reply .gitmodules:1]: ok, reverting

File: lib/Language/Souffle/Interpreted.hs [hunk 7]
```diff
@@ -367,7 +367,7 @@ datalogProgramFile prog datalogDir = do
 
 locateSouffle :: IO (Maybe FilePath)
 locateSouffle = do
-  let locateCmd = (shell "which souffle") { std_out = CreatePipe }
+  let locateCmd = (shell "souffle") { std_out = CreatePipe }
```
Comment (luc-tielen) [inline lib/Language/Souffle/Interpreted.hs:370]:
```suggestion
  let locateCmd = (shell "which souffle") { std_out = CreatePipe }
```

**Review by luc-tielen — APPROVED** [review]
